torch/utils/data/distributed.py

```diff
@@ -65,3 +65,117 @@ def __len__(self):
 
     def set_epoch(self, epoch):
         self.epoch = epoch
+
+class ChunkDataReader(object):
+    r"""Reads a chunk of data given a chunk index.
+
+    A chunk could be full file, section of a large file, folders, URLs, or any other
+    abstraction that allows data to be segmented roughtly the same size.
+
+    As an example, chunking could be used in a scenario where MNIST dataset for training 
+    would be converted in CSV text files and equally splitted in several files,
+    e.g. 60 files with 1000 lines each.  In this case, each worker would read
+    a text file with all 1000 images.
+
+    Another example would be chunks as individual folders, containing multiple audio files.
+    In this scenario, each worker would have access to different folders, 
+    where binary files would be read from the filesystem.
+
+    A chunk doesn necessarily need to rely in multiple files to split data laoding across workers.
+    It is possible to extract chunks from a single large dataset file, where each worker would
+    seek to specific positions to load data.
+
+    The reading logic must be implemented inside `__call__` method
+    """
+
+    def __init__(self):
+        pass
+
+    def __call__(self, idx):
+        r"""Returns `list(samples)`
+
+        If `samples` contain tensors, numpy arrays, numbers,
+        dicts or lists, the default collate function can be implicitly used.
+        Otherwise, a custom `collate_fn` must be provided to `DataLoader`
+
+        When no more data is available, `StopIteration` must be raised.
+        """
+        raise NotImplementedError
+
+class DistributedChunkSampler(Sampler):
+    r"""This sampler introduces distributed sampling without padding and dataset dependency.
+
+    Python DataLoader uses multi-processing instead of
+    multi-threading for parallelism, therefore, each worker
+    is a separate process with an identical copy of sampler.
+    Because of that, on a distributed environment, different processes
+    could read the same portion of the dataset. This sampler uses strides,
+    based on the global worker rank concept, to coordinate parallel reading of chunks of data.
+
+    Each instance of `DistributedChunkSampler` must be configured with a different
+    `global_worker_rank` (aka strides), so that chunk sampling happens in a round-robin
+    fashion on each worker, e.g. rank0, rank1, ..., rankN, rank0, rank1, ..., rankN, ...
+
+    For example, assume 2 workers reading the same `ChunkDataset` dataset.
+    Each worker process needs to configure their `DistributedChunkSampler`
+    so that one of them reads all even batches (e.g. `global_worker_rank` 0)
+    while the other process reads all odd batches (e.g. `global_worker_rank` 1).
+
+    If :attr:`shuffle` is ``False``, `SequentialSampler` will be used to select
+    chunk indicies; ``RandomSampler`` will be used otherwise.
+
+    Args:
+        num_replicas (int): Number of workers participating in the sampling.
```

Comment:
Total number of dataloader workers participating in data reading. Ths is equal to number distributed processes * number of dataloader workers per such process.

Reply:
fixed